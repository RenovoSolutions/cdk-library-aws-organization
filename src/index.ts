import * as path from 'path';
import {
  custom_resources,
  aws_lambda as lambda,
  aws_iam as iam,
  CustomResource,
  aws_logs as logs,
  Duration,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * The supported OrgObject types
 */
export enum OrgObjectTypes {
  OU = 'ou',
  ACCOUNT = 'account',
}

/**
 * The properties of an OU
 */
export interface OUProps {
  /**
   * The name of the OU
   */
  readonly name: string;
  /**
   * Whether or not to import an existing OU if the new OU is a duplicate. If this is false and the OU already exists an error will be thrown.
   *
   * @default false
   */
  readonly importOnDuplicate?: boolean;
  /**
    * Whether or not to merge an OU with a duplicate when an OU is moved between parent OUs.
    * If this is false and the OU already exists an error will be thrown.
    * If this is true and the OU already exists the accounts in the OU will be moved to the existing OU
    * and the duplicate, now empty, OU will be deleted.
    *
    * @default false
    */
  readonly allowMergeOnMove?: boolean;
  /**
    * Whether or not a missing OU should be recreated during an update.
    * If this is false and the OU does not exist an error will be thrown when you try to update it.
    */
  readonly allowRecreateOnUpdate?: boolean;
};

/**
 * The properties of an Account
 */
export interface AccountProps {
  /**
   * The name of the account
   */
  readonly name: string;
  /**
   * The email address of the account. Most be unique.
   */
  readonly email: string;
};

/**
 * The structure of an OrgObject
 */
export interface OrgObject {
  /**
   * The unique id of the OrgObject. This is used as the unique identifier when instantiating a construct object.
   * This is important for the CDK to be able to maintain a reference for the object when utilizing
   * the processOrgObj function rather then using the name property of an object which could change.
   * If the id changes the CDK treats this as a new construct and will create a new construct resource and
   * destroy the old one.
   *
   * Not strictly required but useful when using the processOrgObj function. If the id is not provided
   * the name property will be used as the id in processOrgObj.
   *
   * You can create a unique id however you like. A bash example is provided.
   * @example
   * echo "ou-$( echo $RANDOM | md5sum | head -c 8 )"
   */
  readonly id?: string;
  /**
   * The org object properties.
   */
  readonly properties: OUProps | AccountProps;
  /**
   * The type of the org object.
   */
  readonly type: OrgObjectTypes;
  /**
   * Other org objects that are children of this org object.
   */
  readonly children: OrgObject[];
}

/**
 * @function processOrgObj
 * Function to process an OrgObject and create the corresponding AWS resources
 *
 * @param {Construct} this The construct resources will be added to.
 * @param {custom_resources.Provider} provider The custom resource provider the custom resources will utilized to create the resources.
 * @param {OrgObject} obj The OrgObject to process.
 * @param {string | OrganizationOU} parent The parent of the OrgObject. This is either a string, like for the org root, or an OrganizationOU object from the same stack.
 */
export function processOrgObj(this: Construct, provider: custom_resources.Provider, obj: OrgObject, parent: string | OrganizationOU) {
  if (obj.type === OrgObjectTypes.OU) {
    const parentStr = parent instanceof OrganizationOU ? parent.resource.ref : parent;

    let props: OUProps = obj.properties;
    let id = obj.id ?? obj.properties.name;

    const ou = new OrganizationOU(this, id, {
      provider,
      parent: parentStr,
      name: props.name,
      importOnDuplicate: props.importOnDuplicate,
      allowMergeOnMove: props.allowMergeOnMove,
      allowRecreateOnUpdate: props.allowRecreateOnUpdate,
    });

    obj.children.forEach(child => {
      processOrgObj.call(this, provider, child, ou);
    });
  }
}

/**
 * The properties for the OU custom resource provider.
 */
export interface OrganizationOUProviderProps {
  /**
   * The role the custom resource should use for taking actions on OUs if one is not provided one will be created automatically
   */
  readonly role?: iam.IRole;
}

/**
 * The provider for OU custom resources
 *
 * This creates a lambda function that handles custom resource requests for creating/updating/deleting OUs.
 */
export class OrganizationOUProvider extends Construct {

  public readonly provider: custom_resources.Provider;

  constructor(scope: Construct, id: string, props: OrganizationOUProviderProps) {
    super(scope, id);

    let role: iam.IRole;

    if (!props.role) {
      role = new iam.Role(this, 'role', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      });

      let policy = new iam.ManagedPolicy(this, 'policy', {
        statements: [
          new iam.PolicyStatement({
            actions: [
              'organizations:UpdateOrganizationalUnit',
              'organizations:DeleteOrganizationalUnit',
              'organizations:ListOrganizationalUnitsForParent',
              'organizations:CreateOrganizationalUnit',
            ],
            resources: ['*'],
          }),
        ],
      });

      role.addManagedPolicy(policy);
      role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'));
    } else {
      role = props.role;
    }

    const handlersPath = path.join(__dirname, '../handlers');

    const onEvent = new lambda.Function(this, 'handler', {
      runtime: lambda.Runtime.PYTHON_3_9,
      code: lambda.Code.fromAsset(handlersPath + '/ou'),
      handler: 'index.on_event',
      timeout: Duration.seconds(10),
      role,
    });

    this.provider = new custom_resources.Provider(this, 'provider', {
      onEventHandler: onEvent,
      logRetention: logs.RetentionDays.ONE_DAY,
    });
  }
}

/**
 * The properties of an OrganizationOU custom resource.
 */
export interface OUResourceProps extends OUProps {
  /**
   * The parent OU id
   */
  readonly parent: string | OrganizationOU;
  /**
   * The provider to use for the custom resource that will create the OU. You can create a provider with the OrganizationOuProvider class
   */
  readonly provider: custom_resources.Provider;
}

/**
 * The construct to create or update an Organization OU
 *
 * This relies on the custom resource provider OrganizationOUProvider.
*/
export class OrganizationOU extends Construct {

  public readonly resource: CustomResource;

  constructor(scope: Construct, id: string, props: OUResourceProps) {
    super(scope, id);

    const importOnDuplicate = props.importOnDuplicate ?? false;
    const allowMergeOnMove = props.allowMergeOnMove ?? false;
    const allowRecreateOnUpdate = props.allowRecreateOnUpdate ?? false;

    const parentStr = props.parent instanceof OrganizationOU ? props.parent.resource.ref : props.parent;

    this.resource = new CustomResource(this, 'ou', {
      serviceToken: props.provider.serviceToken,
      properties: {
        Parent: parentStr,
        Name: props.name,
        ImportOnDuplicate: importOnDuplicate,
        AllowMergeOnMove: allowMergeOnMove,
        AllowRecreateOnUpdate: allowRecreateOnUpdate,
      },
    });

    this.resource.node.addDependency(props.provider);
    if (props.parent instanceof OrganizationOU) {
      this.resource.node.addDependency(props.parent);
    };
  }
}

// export interface OrganizationAccountProps {
//   /**
//    * The name of the account
//    */
//    readonly name: string;
//   /**
//    * The email for the account
//    */
//    readonly email: string;
//   /**
//   * The OU id for the account. This property is always preferred if id and name are both given. Defaults to the root
//   */
//    readonly ou_id?: string;
//   /**
//    * The OU name for the account. This property is ignored if the ou id is given.
//    */
//    readonly ou_name?: string;
// }

// export class OrganizationAccount extends Construct {
//   constructor(scope: Construct, id: string, props: OrganizationAccountProps) {
//     super(scope, id);

//     // code
//   }
// }
