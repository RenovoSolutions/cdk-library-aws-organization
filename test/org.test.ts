import { Stack, App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { OrganizationOU, OrganizationOUProvider } from '../src/index';

test('Snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  new OrganizationOU(stack, 'ou', {
    name: 'test',
    parentId: 'r-1234',
    provider: new OrganizationOUProvider(stack, 'provider', {}).provider,
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
