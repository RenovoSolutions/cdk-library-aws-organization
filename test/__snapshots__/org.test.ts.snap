// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c13434f8f1aa2ea30fa577b2feb208a41368b11787b752e10bfc71fe8eb919d5.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationAccountProviderhandlerEAF7C45E": Object {
      "DependsOn": Array [
        "OrganizationAccountProviderroleFD0EE157",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0cd5206a8d9b08f8135770a304f9c28303ba9457bac2474306d59aaa8c47c6c1.zip",
        },
        "Handler": "index.on_event",
        "Role": Object {
          "Fn::GetAtt": Array [
            "OrganizationAccountProviderroleFD0EE157",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OrganizationAccountProviderpolicy92ABFDF8": Object {
      "Properties": Object {
        "Description": "",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "organizations:ListOrganizationalUnitsForParent",
                "organizations:ListAccountsForParent",
                "organizations:ListRoots",
                "organizations:MoveAccount",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "OrganizationAccountProviderproviderframeworkonEventA1FDA928": Object {
      "DependsOn": Array [
        "OrganizationAccountProviderproviderframeworkonEventServiceRoleDefaultPolicy05A6AEB5",
        "OrganizationAccountProviderproviderframeworkonEventServiceRole082F1D31",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (TestStack/OrganizationAccountProvider/provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "OrganizationAccountProviderhandlerEAF7C45E",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "OrganizationAccountProviderproviderframeworkonEventServiceRole082F1D31",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OrganizationAccountProviderproviderframeworkonEventLogRetentionB1D63354": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "OrganizationAccountProviderproviderframeworkonEventA1FDA928",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "OrganizationAccountProviderproviderframeworkonEventServiceRole082F1D31": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OrganizationAccountProviderproviderframeworkonEventServiceRoleDefaultPolicy05A6AEB5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "OrganizationAccountProviderhandlerEAF7C45E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationAccountProviderproviderframeworkonEventServiceRoleDefaultPolicy05A6AEB5",
        "Roles": Array [
          Object {
            "Ref": "OrganizationAccountProviderproviderframeworkonEventServiceRole082F1D31",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationAccountProviderroleFD0EE157": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Ref": "OrganizationAccountProviderpolicy92ABFDF8",
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OrganizationOUProviderhandler080F4AD3": Object {
      "DependsOn": Array [
        "OrganizationOUProviderroleA45081DC",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "686a4d7c54a1ddeb61aa6fd3ee2dd61897a21162d36c77dbb12e4207f064b44b.zip",
        },
        "Handler": "index.on_event",
        "Role": Object {
          "Fn::GetAtt": Array [
            "OrganizationOUProviderroleA45081DC",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OrganizationOUProviderpolicyC8E987A8": Object {
      "Properties": Object {
        "Description": "",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "organizations:UpdateOrganizationalUnit",
                "organizations:DeleteOrganizationalUnit",
                "organizations:ListOrganizationalUnitsForParent",
                "organizations:CreateOrganizationalUnit",
                "organizations:DescribeAccount",
                "organizations:DescribeCreateAccountStatus",
                "organizations:CreateAccount",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "OrganizationOUProviderproviderframeworkonEvent455BD921": Object {
      "DependsOn": Array [
        "OrganizationOUProviderproviderframeworkonEventServiceRoleDefaultPolicy1A8B9C1C",
        "OrganizationOUProviderproviderframeworkonEventServiceRole32FF7533",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (TestStack/OrganizationOUProvider/provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "OrganizationOUProviderhandler080F4AD3",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "OrganizationOUProviderproviderframeworkonEventServiceRole32FF7533",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "OrganizationOUProviderproviderframeworkonEventLogRetention49235E57": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "OrganizationOUProviderproviderframeworkonEvent455BD921",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "OrganizationOUProviderproviderframeworkonEventServiceRole32FF7533": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OrganizationOUProviderproviderframeworkonEventServiceRoleDefaultPolicy1A8B9C1C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "OrganizationOUProviderhandler080F4AD3",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationOUProviderproviderframeworkonEventServiceRoleDefaultPolicy1A8B9C1C",
        "Roles": Array [
          Object {
            "Ref": "OrganizationOUProviderproviderframeworkonEventServiceRole32FF7533",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationOUProviderroleA45081DC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Ref": "OrganizationOUProviderpolicyC8E987A8",
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestAccounttestexamplecomouCFF56022": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationAccountProviderproviderframeworkonEventLogRetentionB1D63354",
        "OrganizationAccountProviderproviderframeworkonEventA1FDA928",
        "OrganizationAccountProviderproviderframeworkonEventServiceRoleDefaultPolicy05A6AEB5",
        "OrganizationAccountProviderproviderframeworkonEventServiceRole082F1D31",
        "oudf425b37ou574381B2",
      ],
      "Properties": Object {
        "AllowMove": false,
        "DisableDelete": false,
        "ImportOnDuplicate": false,
        "Name": "TestAccount",
        "Parent": Object {
          "Ref": "oudf425b37ou574381B2",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationAccountProviderproviderframeworkonEventA1FDA928",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "ou301416f8ouDD4B1D0E": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationOUProviderproviderframeworkonEventLogRetention49235E57",
        "OrganizationOUProviderproviderframeworkonEvent455BD921",
        "OrganizationOUProviderproviderframeworkonEventServiceRoleDefaultPolicy1A8B9C1C",
        "OrganizationOUProviderproviderframeworkonEventServiceRole32FF7533",
      ],
      "Properties": Object {
        "AllowRecreateOnUpdate": false,
        "ImportOnDuplicate": false,
        "Name": "TestOU",
        "Parent": "r-a1b2",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationOUProviderproviderframeworkonEvent455BD921",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "oudf425b37ou574381B2": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationOUProviderproviderframeworkonEventLogRetention49235E57",
        "OrganizationOUProviderproviderframeworkonEvent455BD921",
        "OrganizationOUProviderproviderframeworkonEventServiceRoleDefaultPolicy1A8B9C1C",
        "OrganizationOUProviderproviderframeworkonEventServiceRole32FF7533",
      ],
      "Properties": Object {
        "AllowRecreateOnUpdate": false,
        "ImportOnDuplicate": false,
        "Name": "TestOUChild",
        "Parent": Object {
          "Ref": "ou301416f8ouDD4B1D0E",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationOUProviderproviderframeworkonEvent455BD921",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
